File: sora-editor/src/main/java/io/github/rosemoe/sora/text/ProblemMarker.java
package io.github.rosemoe.sora.text;

import com.pranav.analyzer.java.JavacAnalyzer;
import com.pranav.common.util.ConcurrentUtil;
import com.pranav.common.util.DiagnosticWrapper;
import com.pranav.common.Indexer;
import com.pranav.common.util.FileUtil;

import io.github.rosemoe.sora.text.LineNumberCalculator;
import io.github.rosemoe.sora.widget.CodeEditor;
import io.github.rosemoe.sora.util.HighlightUtil;

import java.io.IOException;

public class ProblemMarker implements ContentListener {

    private CodeEditor editor;
    private JavacAnalyzer analyzer;

    public ProblemMarker(CodeEditor editor) {
        this.editor = editor;
        this.analyzer = new JavacAnalyzer(editor.getContext());
    }

    @Override
    public void beforeReplace(Content content) {}
    
    @Override
    public void afterInsert(
                Content content,
                int startLine,
                int startColumn,
                int endLine,
                int endColumn,
                CharSequence insertedContent
    ) {
      run(content);
    }
    
    @Override
    public void afterDelete(
                Content content,
                int startLine,
                int startColumn,
                int endLine,
                int endColumn,
                CharSequence deletedContent
    ) {
      run(content);
    }
    
    
    private void run(Content content) {
        ConcurrentUtil.inParallel(
                () -> {
                    if (!analyzer.isFirstRun()) {
                        analyzer.reset();
                    }
                    try {
                        FileUtil.writeFile(new Indexer("editor").getString("currentFile"), content.toString());
                        analyzer.analyze();
                    } catch (Exception ignored) {
                        // we shouldn't disturb the user for some issues
                    }
                    HighlightUtil.clearDiagnostics(editor.getStyles());
                    HighlightUtil.markDiagnostics(
                            editor, analyzer.getDiagnostics(), editor.getStyles());
                });
    }

    private void setLineAndColumn(DiagnosticWrapper diagnostic) {
        try {
            // Calculate and update the start and end line number and columns
            var startCalculator = new LineNumberCalculator(editor.getText().toString());
            startCalculator.update((int) diagnostic.getLineNumber());
            diagnostic.setStartLine((int) diagnostic.getLineNumber());
            diagnostic.setStartColumn(startCalculator.getColumn());
            var endCalculator = new LineNumberCalculator(editor.getText().toString());
            endCalculator.update((int) diagnostic.getLineNumber());
            diagnostic.setEndLine((int) diagnostic.getLineNumber());
            diagnostic.setEndColumn(endCalculator.getColumn());
        } catch (IndexOutOfBoundsException ignored) {
            // unknown index, dont update line numbers
        }
    }
}

File: android-compiler/src/main/java/org/cosmic/ide/ProblemMarker.kt
package org.cosmic.ide

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import io.github.rosemoe.sora.lang.diagnostic.DiagnosticsContainer
import io.github.rosemoe.sora.text.Content
import io.github.rosemoe.sora.text.ContentListener
import io.github.rosemoe.sora.widget.CodeEditor
import org.cosmic.ide.analyzer.java.JavacAnalyzer
import org.cosmic.ide.common.util.CoroutineUtil
import org.cosmic.ide.common.util.FileUtil
import org.cosmic.ide.project.JavaProject
import java.io.File

class ProblemMarker(
    private val context: Context,
    private val editor: CodeEditor,
    private val file: File,
    private val project: JavaProject
) : ContentListener {

    private val analyzer: JavacAnalyzer by lazy { JavacAnalyzer(context, project) }
    private val diagnostics = DiagnosticsContainer()
    private val TAG = "ProblemMarker"

    init {
        analyze(editor.text)
    }

    override fun beforeReplace(content: Content) {}

    override fun afterInsert(
        content: Content,
        startLine: Int,
        startColumn: Int,
        endLine: Int,
        endColumn: Int,
        insertedContent: CharSequence
    ) {
        analyze(content)
    }

    override fun afterDelete(
        content: Content,
        startLine: Int,
        startColumn: Int,
        endLine: Int,
        endColumn: Int,
        deletedContent: CharSequence
    ) {
        analyze(content)
    }

    private fun analyze(content: Content) {
        val code = content.toString()
        FileUtil.writeFile(file.absolutePath, code)
        CoroutineUtil.inParallel {
            if (!analyzer.isFirstRun()) {
                analyzer.reset()
            }
            try {
                if (!(file.extension.equals("java"))) {
                    Handler(Looper.getMainLooper()).post {
                        editor.diagnostics = DiagnosticsContainer()
                    }
                    return@inParallel
                }
                analyzer.analyze()
                diagnostics.reset()
                diagnostics.addDiagnostics(analyzer.getDiagnostics())
                Handler(Looper.getMainLooper()).post {
                    editor.diagnostics = diagnostics
                }
            } catch (e: Exception) {
                Log.d(TAG, "Error while marking diagnostics.", e)
            }
        }
    }
}
